test: add langchain-google-genai integration tests for RAI features (…

Ranga-Prasath-22 opened the pull request:
…#168)

This test suite verifies the integration and functionality of the ChatGoogleGenerativeAI model, including initialization, invocation, error handling, and support for model variants.

## Purpose

-   Clear purpose of this PR

## Proposed Changes

What does this PR add, remove or fix?

## Issues

-   Links to relevant issues

## Testing

-   How was it tested, what were the results?

Ranga-Prasath-22 committed:
test: add langchain-google-genai integration tests for RAI features (#168)

This test suite verifies the integration and functionality of the ChatGoogleGenerativeAI model, including initialization, invocation, error handling, and support for model variants.

maciejmajek commented:
Thank you for the contribution! I appreciate the effort you put into adding tests for Google GenAI integration.

A few points to note:

* The PR does not add the `langchain-google-genai` dependency, so these tests won’t run out of the box.
* The added tests are heavily mocked. While they might make sense for testing the LangChain package itself, they don’t really test RAI, which uses LLMs in a less conventional way. Fully mocked tests like these don’t provide much value here.
* For RAI, testing LLM integrations works better through manual testing. Before configuring the model in `config.toml`, the initialization module should be updated to support the new LLM vendor. Then you can manually test by running the demos to verify everything works as expected.

Just so you know, our current contributor @sachinkum0009 is also working on Google GenAI integration. You could coordinate to pick up the work together or test the capabilities separately.

Thanks again!

sachinkum0009 commented:
Thanks for the PR and for the heads-up! I actually haven't started working on this specific issue yet. If you are already mid-way through, feel free to take the lead. Otherwise, I’m happy to discuss how we can divide the tasks so we aren't duplicating effort. What do you think?

Ranga-Prasath-22 commented:
@maciejmajek Thank you for the feedback! I'll update the PR to include the `langchain-google-genai` dependency and add support for the Google LLM vendor in the initialization module. I'll also work on making the tests more meaningful for RAI.

@sachinkum0009 Thanks! I'm already partway through these changes, so I'll go ahead and finish the core setup. We can definitely coordinate on further features once the basic integration is merged!

Ranga-Prasath-22 committed:
feat: add langchain-google-genai integration for Google Gemini models

- Add langchain-google-genai dependency to pyproject.toml
- Add GoogleConfig dataclass and vendor support in model_initialization.py
- Add [google] section to config template with gemini-3-flash/gemini-3-pro
- Add unit tests for RAI framework integration logic

Closes #168

Ranga-Prasath-22 committed:
Merge branch 'RobotecAI:main' into main

Ranga-Prasath-22 committed:
Merge pull request #2 from Ranga-Prasath-22/main

Syncing patch-1 with main (latest fixes and upstream changes)

Ranga-Prasath-22 commented:
@maciejmajek @sachinkum0009 Thank you for the guidance! I've updated the PR with the following changes:
- Added `langchain-google-genai` as a dependency in `pyproject.toml`.
- Updated `model_initialization.py` to support the Google LLM vendor.
- Added the `[google]` configuration section to `config_initialization.py` and the `config.toml` template.
- Added comprehensive unit tests for the initialization logic and integrated the Google GenAI model.

Please let me know if there's anything else I should address. This is my first open-source contribution, so I appreciate your patience!

Ranga-Prasath-22 committed:
fix: make langchain-google-genai a lazy import and bump version to 2.7.0

- Remove top-level import of langchain_google_genai in model_initialization.py
- Update return type annotations to use Any for lazy-imported types
- Add try/except skip in test_langchain_google_genai.py for optional dependency
- Bump version from 2.6.0 to 2.7.0 for new feature addition

Fixes CI build errors caused by missing optional dependency during test collection

Ranga-Prasath-22 committed:
style: fix formatting in test_google_genai_initialization.py

sachinkum0009 commented on src/rai_core/rai/initialization/model_initialization.py:
```diff
@@ -130,7 +138,7 @@ def get_llm_model(
     config_path: Optional[str] = None,
     **kwargs: Any,
-) -> ChatOpenAI | ChatBedrock | ChatOllama:
+) -> ChatOpenAI | ChatBedrock | ChatOllama | Any:
```
use `ChatGoogleGenerativeAI` instead of `Any`

use this to import
from langchain_google_genai import ChatGoogleGenerativeAI

sachinkum0009 commented on src/rai_core/rai/initialization/model_initialization.py:
```diff
@@ -166,7 +179,7 @@ def get_llm_model_direct(
     config_path: Optional[str] = None,
     **kwargs: Any,
-) -> ChatOpenAI | ChatBedrock | ChatOllama:
+) -> ChatOpenAI | ChatBedrock | ChatOllama | Any:
```
Same as above.

sachinkum0009 commented on tests/agents/langchain/test_langchain_google_genai.py:
I think this test is not required. It doesn't align with agents. @maciejmajek  what are your thoughts on this?

maciejmajek replied:
Agreed. These are more of google-genai tests and are not relevant to rai

sachinkum0009 commented on tests/initialization/test_google_genai_initialization.py:
Similarly, this test is also not required. 
I suggest to update the `test_model_initialization.py` file to include the google genai config and then add the unit test to validate the model initialization from config.
@maciejmajek What are your suggestions for this?

sachinkum0009 commented:
Thanks @Ranga-Prasath-22, so far it looks good. I have suggested some changes.

sachinkum0009 commented on src/rai_core/pyproject.toml:
```diff
@@ -26,6 +26,7 @@ langchain = "*"
 langchain-aws = "*"
 langchain-openai = "*"
 langchain-ollama = "^0.3.4"
+langchain-google-genai = "*"
```
@maciejmajek Is it fine to update `langchain-core` to latest version, as it is required to install `langchain-google-genai`?

maciejmajek replied:
Unfortunately, due to numpy 1.x requirement of ROS 2 jazzy and humble we are stuck with the langchain-core 0.X version

Ranga-Prasath-22 commented:
Thanks @sachinkum0009! I'm looking forward to your suggestions. Once you submit your review, I'll work on addressing those changes. I'm ready to update the branch with the main branch whenever you're done with your review.

Ranga-Prasath-22 commented:
Hey @sachinkum0009 @maciejmajek, thanks for the feedback! I've got all the changes sorted:

- Updated the return types to use `ChatGoogleGenerativeAI` instead of `Any` (with the proper import from `langchain_google_genai`)
- Removed the agent test file since it doesn't really fit with RAI
- Deleted the separate initialization test and moved everything into `test_model_initialization.py`
- Kept `langchain-core` on the 0.x version to stay compatible with ROS2's numpy 1.x requirement

Should be good to review now!

sachinkum0009 commented:
LGTM!

maciejmajek reviewed (approved):
LGTM! Thank you @Ranga-Prasath-22 and @sachinkum0009!

maciejmajek commented:
@Ranga-Prasath-22 as a final step, could you please resolve the conflicts and bump the rai-core minor version?
So whatever version it is now e.g. 2.x.y bump the x by 1 and zero the y. Thanks!

Ranga-Prasath-22 committed:
Merge upstream/main: resolve conflicts and bump rai-core to 2.9.0

Ranga-Prasath-22 commented:
Hi @maciejmajek @sachinkum0009 — all done! ✅

Merged upstream `main`, resolved the conflicts, and bumped `rai-core` to **2.9.0** (minor +1, patch zeroed).

Thank you both for the guidance and patience throughout this PR. This has been my first open-source contribution, and I truly appreciate the opportunity to contribute to RAI. Learned a lot from the review process! 🙏

maciejmajek commented:
@Ranga-Prasath-22 hi, I consider this PR ready to be merged, just waiting for our internal CI system to come back online. Sorry for the wait!

KE7 commented:
Would adding `litellm` be easier than additional langchain dependencies?

Ranga-Prasath-22 commented:
Hi @KE7, thanks for the suggestion! I considered litellm, but since RAI already uses LangChain as its core LLM framework (with langchain-openai already being a dependency), using langchain-google-genai keeps things consistent with the existing architecture and avoids introducing a separate abstraction layer. It also allows the tests to directly validate how Google's Gemini models work within RAI's LangChain-based pipeline. That said, I'm happy to explore the litellm approach if the maintainers feel it would be a better fit going forward!

maciejmajek commented:
Hi @Ranga-Prasath-22, are you able to reproduce the CI errors from this branch locally?

@Ranga-Prasath-22 @KE7 Regarding litellm, I think it would be a cool addition. While most of the code is already bounded to langchain, litellm can be used to setup the model for langchain which is very cool.

@KE7 is that something that you would like to see in rai? I'm open to it.

maciejmajek committed:
Merge branch 'main' into Ranga-Prasath-22-patch-1

Ranga-Prasath-22 commented:
Hi @maciejmajek, thanks for merging main! I'll pull the latest changes and try to reproduce the CI errors locally. Just to clarify — the failing checks are the build-and-test-ros2 jobs (humble & jazzy). I don't currently have a ROS2 environment set up on my machine. Would you prefer I set one up to test, or could these failures be related to the internal CI issues you mentioned earlier? Happy to investigate further either way!

maciejmajek committed:
chore: update poetry.lock for dependency changes and add new models in tests

- Updated Poetry version from 2.1.4 to 2.3.2.
- Added new package dependencies: `filetype` and `google-ai-generativelanguage`.
- Updated `google-api-core`, `google-auth`, and `grpcio` dependencies.
- Modified test configurations to include new Google models: `gemini-3-flash`, `gemini-3-pro`, and `text-embedding-004`.
- Removed obsolete test file `test_langchain_google_genai.py`.

wip

maciejmajek commented:
@Ranga-Prasath-22 I took over to debug some things. Please cherry pick [ef12a37](https://github.com/RobotecAI/rai/pull/767/commits/ef12a37e47e3ad5956bf638fb99a567dd203223e)
Then we will finally merge it. Thanks!

Ranga-Prasath-22 commented:
Done! Cherry-picked ef12a37 and pushed. CI should be re-running now.

KE7 commented:
@maciejmajek I think from a lightness perspective, litellm would just make rai a smaller package to install. Langchain is quite heavy. As you can see, just adding gemini here required a new dependency whereas something like litellm would just be a string change on the user's API request and then you're done.
